# Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en 100 gramos, se deben tener por lo menos 8 gramos de proteína y 6 gramos de grasa. Así mismo, no se debe tener más de 2 gramos de fibra y 0.4 gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, para simplificar la notación se van a nombrar las siguientes variables: 

$X_1:$ Porcentaje de pollo  
$X_2:$ Porcentaje de carne  
$X_3:$ Porcentaje de cordero  
$X_4:$ Porcentaje de arroz  
$X_5:$ Porcentaje de trigo  
$X_6:$ Porcentaje de gel  

Con los datos, se puede plantear la función objetivo, está dada por la siguiente expresión:

$Minimizar: Costo=$ $0.013X_1 + 0.008X_2 + 0.010X_3 + 0.002X_4 + 0.005X_5 + 0.001X_6$

Las restricciones estarían dadas por el siguiente conjunto de ecuaciones:

$X_1+X_2+X_3+X_4+X_5+X_6=100$  

$(10.0 X_1+ 20.0 X_2+ 15.0 X_3+ 00.0 X_4+ 04.0 X_5+ 00.0 X_6)/100 \geq 8.0$  

$(08.0 X_1+ 10.0 X_2+ 11.0 X_3+ 01.0 X_4+ 01.0 X_5+ 00.0 X_6)/100 \geq 6.0$  

$(00.1 X_1+ 00.5 X_2+ 00.5 X_3+ 10.0 X_4+ 15.0 X_5+ 00.0 X_6)/100 \leq 2.0$  

$(00.2 X_1+ 00.5 X_2+ 00.7 X_3+ 00.2 X_4+ 00.8 X_5+ 00.0 X_6)/100 \leq 0.4$  

La primer condición asegura que la cantidad de productos que se usará cumple con el 100%. Las siguientes sólo siguen los lineamientos planteados para cumplir con los requisitos nutrimentales. 

In [7]:
from pulp import *

#Se genera la variable del problema,
prob=LpProblem("Costo mínimo de la dieta", LpMinimize)

#Se definen las variables que serán utilizadas
X_1=LpVariable("X1: Porcentaje de pollo", lowBound=0)
X_2=LpVariable("X2: Porcentaje de carne", lowBound=0)
X_3=LpVariable("X3: Porcentaje de cordero", lowBound=0)
X_4=LpVariable("X4: Porcentaje de arroz", lowBound=0)
X_5=LpVariable("X5: Porcentaje de trigo", lowBound=0)
X_6=LpVariable("X6: Porcentaje de gel", lowBound=0)

#Se crea la función objetivo:
Funcion_objetivo= 0.013*X_1 + 0.008*X_2 + 0.010*X_3 + 0.002*X_4 + 0.005*X_5 + 0.001*X_6

#Se crean las resteicciones

restriccion1=X_1+X_2+X_3+X_4+X_5+X_6 == 100
restriccion2=(10.0*X_1+ 20.0*X_2+ 15.0*X_3+ 00.0*X_4+ 04.0*X_5+ 00.0*X_6)/100 >= 8.0  
restriccion3=(08.0*X_1+ 10.0*X_2+ 11.0*X_3+ 01.0*X_4+ 01.0*X_5+ 00.0*X_6)/100  >= 6.0
restriccion4=(00.1*X_1+ 00.5*X_2+ 00.5*X_3+ 10.0*X_4+ 15.0*X_5+ 00.0*X_6)/100  <= 2.0
restriccion5=(00.2*X_1+ 00.5*X_2+ 00.7*X_3+ 00.2*X_4+ 00.8*X_5+ 00.0*X_6)/100  <= 0.4

#Se agrega la funcion objetivo y las restricciones al problema
prob += Funcion_objetivo
prob += restriccion1
prob += restriccion2
prob += restriccion3
prob += restriccion4
prob += restriccion5

#Se soluciona el problema:
prob.solve()

#Resultados impresos:
print ("Status:", LpStatus[prob.status])
for v in prob.variables():
    print(v.name, "=", v.varValue,"%")
print("Valor Óptimo = $", value(prob.objective))



Status: Optimal
X1:_Porcentaje_de_pollo = 0.0 %
X2:_Porcentaje_de_carne = 60.0 %
X3:_Porcentaje_de_cordero = 0.0 %
X4:_Porcentaje_de_arroz = 0.0 %
X5:_Porcentaje_de_trigo = 0.0 %
X6:_Porcentaje_de_gel = 40.0 %
Valor Óptimo = $ 0.52


Se puede observar que con un 40% de gel y 60% de carne se estan cumpliendo los requerimientos y asi mismo, se tiene el costo mínimo que es de $\$.52$. 